Create device service feature.

abienias97 opened the pull request.

abienias97 committed:
Create device service feature.

kowallus commented on src/main/java/edu/kis/powp/jobs2d/drivers/adapter/monitor/DeviceMonitorDriver.java:
```diff
@@ -23,10 +26,18 @@ public void setPosition(int x, int y) {
+        if (DeviceMonitor.getInstance().checkServiceDistanceThreshold(2000.0)) {
+            logger.info("Service threshold exceeded");
+            return;
+        }
```
This is roughly hardcoded solution. The number is fixed, and there is no OCP for implementing reaction to "threshold exceeded" event. 
The soultion should be configurable (at least be parameters injection) and OCP. Maybe use the observer pattern to enable triggering various reactions (e.g., logger notification)

abienias97 replied:
Added more OCP and configuring.

kowallus commented on src/main/java/edu/kis/powp/jobs2d/drivers/adapter/monitor/DeviceMonitorDriver.java:
```diff
@@ -23,10 +26,18 @@ public void setPosition(int x, int y) {
+    public void refill() {
+        DeviceMonitor.getInstance().refill();
+    }
+
```
The semantics of refilling are not suitable for DeviceMonitorDriver (monitoring seems to be a passive task). Maybe refactor or renaming could make more sense out of it?

abienias97 replied:
Changed semantics and refactored.

kowallus reviewed (changes requested):
More OCP and configurability please.

abienias97 committed:
Refactor to use listeners, rename class to better show its usage.

kowallus commented on src/main/java/edu/kis/powp/jobs2d/drivers/adapter/controller/DeviceControllerDriver.java:
```diff
@@ -0,0 +1,66 @@
+public class DeviceControllerDriver implements DriverDecorator {
+
+    private VisitableJob2dDriver driver;
+    private final List<IThresholdListener> listeners = new ArrayList<>();
```
OK, but the observer pattern is already implemented in this project (Publisher & Subscriber) classes.

kowallus commented on src/main/java/edu/kis/powp/jobs2d/drivers/adapter/controller/DeviceControllerDriver.java:
```diff
@@ -0,0 +1,66 @@
+    private double distanceThreshold = 0;
+
+    public DeviceControllerDriver() {
+        addListener(new LoggingThresholdListener());
```
SRP - listeners should be configured elsewhere.

kowallus commented on src/main/java/edu/kis/powp/jobs2d/drivers/adapter/controller/DeviceControllerDriver.java:
```diff
@@ -0,0 +1,66 @@
+        addListener(new LoggingThresholdListener());
+    }
+
+    public void setDistanceThreshold(double distanceThreshold) {
+        this.distanceThreshold = distanceThreshold;
+    }
```
Unfortunately, we are still limited to a single distance threshold.

kowallus commented on src/main/java/edu/kis/powp/jobs2d/drivers/adapter/controller/DeviceControllerDriver.java:
```diff
@@ -0,0 +1,66 @@
+
+    @Override
+    public void operateTo(int x, int y) {
+        if (DeviceController.getInstance().checkServiceDistanceThreshold(distanceThreshold)) {
```
If distance is maintained by DeviceController maybe observer pattern should be located in that class?

kowallus commented on src/main/java/edu/kis/powp/jobs2d/drivers/singleton/DeviceController.java:
```diff
@@ -4,15 +4,15 @@
 import java.util.logging.Logger;
 
-public final class DeviceMonitor implements RefillableDevice {
+public final class DeviceController implements RefillableDevice {
```
The name does not seem suitable at all.

kowallus reviewed (commented):
Currently, I propose grade 3 for the project part.

Kmynek committed:
Fixed bugs for PR/25

Refactor DeviceController to DeviceManager and implement dynamic distance thresholds

- Renamed DeviceController to DeviceManager for clarity.
- Centralized observer (threshold listeners) handling in DeviceManager.
- Replaced single distance threshold with a dynamic list of thresholds.
- Updated checkServiceDistanceThreshold() to iterate over the list and notify observers.
- In DeviceControllerDriver, replaced setDistanceThreshold() with addDistanceThreshold() delegating to DeviceManager.

kowallus commented on src/main/java/edu/kis/powp/jobs2d/drivers/adapter/controller/DeviceControllerDriver.java:
```diff
@@ -2,52 +2,46 @@
 
     public DeviceControllerDriver() {
-        addListener(new LoggingThresholdListener());
+        DeviceManager.getInstance().addThresholdListener(new LoggingThresholdListener());
```
SRP - listeners should be configured elsewhere. They are still configured inside DeviceControllerDriver!
addThresholdListener can be easily called from outside this class.

kowallus commented on src/main/java/edu/kis/powp/jobs2d/drivers/adapter/controller/DeviceControllerDriver.java:
```diff
@@ -2,52 +2,46 @@
     public DeviceControllerDriver(VisitableJob2dDriver driver) {
         this.driver = driver;
-        addListener(new LoggingThresholdListener());
+        DeviceManager.getInstance().addThresholdListener(new LoggingThresholdListener());
```
Remove dependency to LoggingThresholdListener.

kowallus reviewed (changes requested):
Try coping with configuring LoggingThresholdListener independent from DeviceControllerDriver class

Kmynek committed:
LoggingThresholdListener-fix

Moved LoggingThresholdListener setup out of DeviceControllerDriver to keep it focused on controlling the device. Now it’s added in ServicableFeature during init -listener configuration is now  handled separately.
